### PR TITLE
feat(btc-server): pegout tx fee and coin selection improvements

### PR DESCRIPTION
**Pull request by DariusParvin**
## Issue being fixed or feature implemented
This PR helps address fee estimation and pegout reliability issues described in [botanix-labs/botanix#891](https://github.com/botanix-labs/botanix/issues/891), with improvements to fee targeting, change output handling, and dust protection.

## What was done?
- Refactored coin_selection.rs unit tests to support reusable test setup for various pegout scenarios.
- Added helper methods to accurately calculate weight and fee rate for fully signed pegout PSBTs.
- Improved fee targeting so that the final signed PSBT always matches the desired feerate exactly.
- Always include a change output to ensure pegout outputs pay fees.
- Introduced a configurable pegout change output target, defaulting to 50% of pegout value, with min/max bounds to balance liquidity preservation and UTXO bloat:
    - Min: 0.0001 BTC
    - Max: 5% of remaining UTXO value
- Added protection against creating dust outputs after subtracting fees. If dust outputs are removed, fees and change are recalculated to maintain accuracy.

-

## How Has This Been Tested?
- Unit tests for all modified or new methods
- Added fee rate check on e2e_frost integration test

## Breaking Changes

- None

## Checklist

-   [x] I have added or updated relevant unit/integration/functional/e2e tests
-   [ ] I have tested my changes running a local network, and they work as expected
-   [x] I have performed a self-review
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
-   [x] I have made corresponding changes to the documentation if needed
-   [x] I have added assignees and reviewers to this pull request
-   [x] I have added the PR to the project board or linked it to an issue from the board

## Summary by CodeRabbit

* **New Features**
  * Improved coin selection logic for more robust fee and dust handling, with clearer error messages and better output filtering.
  * Added utility functions to estimate the weight and fee rate of fully signed transactions.
  * Introduced new helper functions for generating P2WPKH scriptPubKeys and adding dummy signatures for transaction simulation.
  * Added public constants representing SegWit marker and flag weights for transaction calculations.

* **Bug Fixes**
  * Enhanced handling of outputs that become dust after fee subtraction, reducing the chance of transaction errors.

* **Tests**
  * Added comprehensive unit tests for coin selection, fee calculation, and transaction weight/fee rate estimation.
  * Extended end-to-end tests to verify pegout transaction fee accuracy against expected fee rates.

### CODE REVIEW

**Comment by DariusParvin** [comment]
Since I combined the PRs into one, i'm referencing the commits where I address the feedback for @rwlockbg and @lamafab:

@rwlockbg removing `expect` and returning `Result`:
- [fb7ca9b](https://github.com/botanix-labs/Macbeth/pull/871/commits/fb7ca9b757a1c954991c35c17433384adc5e06e0)
- [6b0939f](https://github.com/botanix-labs/Macbeth/pull/871/commits/6b0939f01d4e0277a78ad31ab28dabf28441916a)

@lamafab 
- estimate weight and fee using dummy signatures [b74f066](https://github.com/botanix-labs/Macbeth/pull/871/commits/b74f066d25dbeb37a10c3601e378bdaaefbda405)
- add expected change amount to coin_selection tests: [489c27e](https://github.com/botanix-labs/Macbeth/pull/871/commits/489c27e635a7303d60ad4785547ed9f7af89c1a7)
- adding fee rate check to e2e test[0a36676](https://github.com/botanix-labs/Macbeth/pull/871/commits/0a3667650dd713a88c5811cf7664febb2ccca841)

File: bin/btc-server/src/test_utils/mod.rs [hunk 1]
```diff
@@ -414,3 +424,27 @@ pub fn store_pending_pegout(db: &database::Db) -> PegoutId {
 
     pegout_id
 }
+
+// Add dummy signatures to a PSBT to help calculate weight and fee rate
+pub fn add_dummy_signatures_to_psbt(mut psbt: Psbt, sighash_type: TapSighashType) -> Psbt {
```
Comment (rwlockbg) [inline bin/btc-server/src/test_utils/mod.rs]: Why not pass in a &mut reference here - I dont see the need of having to move as a mut the entire object ?

File: bin/btc-server/src/wallet/util.rs [hunk 2]
```diff
@@ -44,3 +49,157 @@ pub trait VerifyingKeyExt: Into<frost::VerifyingKey> {
...
+    // calculate the weight of the signatures (assuming all inputs are p2tr)
+    let num_inputs = psbt.inputs.len();
+    let per_input_witness_item_count = Weight::from_wu(1);
+    let total_signature_weight = (TAPROOT_KEYSPEND_SATISFACTION_WEIGHT +
+        per_input_witness_item_count)
+        .checked_mul(num_inputs as u64)
+        .expect("Bitcoin amounts should never overflow u64");
+
+    // total including base weights for segwit transactions
+    unsigned_tx_weight + total_signature_weight + SEGWIT_FLAG_WEIGHT + SEGWIT_MARKER_WEIGHT
```
Comment (rwlockbg) [inline bin/btc-server/src/wallet/util.rs]: I am not an expert here, but can this overflow ?

File: bin/btc-server/src/wallet/util.rs [hunk 3]
```diff
@@ -44,3 +49,157 @@ pub trait VerifyingKeyExt: Into<frost::VerifyingKey> {
...
+pub fn calculate_signed_tx_weight(psbt: &Psbt) -> Weight {
+    let unsigned_tx_weight = psbt.unsigned_tx.weight();
+
+    // calculate the weight of the signatures (assuming all inputs are p2tr)
+    let num_inputs = psbt.inputs.len();
+    let per_input_witness_item_count = Weight::from_wu(1);
+    let total_signature_weight = (TAPROOT_KEYSPEND_SATISFACTION_WEIGHT +
+        per_input_witness_item_count)
+        .checked_mul(num_inputs as u64)
+        .expect("Bitcoin amounts should never overflow u64");
```
Comment (rwlockbg) [inline bin/btc-server/src/wallet/util.rs]: Can we handle this somehow ?

File: bin/btc-server/src/wallet/util.rs [hunk 4]
```diff
@@ -44,3 +49,157 @@ pub trait VerifyingKeyExt: Into<frost::VerifyingKey> {
...
+
+    // total including base weights for segwit transactions
+    unsigned_tx_weight + total_signature_weight + SEGWIT_FLAG_WEIGHT + SEGWIT_MARKER_WEIGHT
+}
+
+// Calculates the fee rate of a PSBT after it has been fully signed with P2TR keyspend inputs.
+pub fn calculate_signed_tx_fee_rate(psbt: &Psbt) -> FeeRate {
+    let tx_weight = calculate_signed_tx_weight(psbt);
+    let absolute_fee = psbt.fee_amount().unwrap();
+    FeeRate::from_sat_per_kwu((absolute_fee.to_sat() * 1000) / tx_weight.to_wu())
```
Comment (rwlockbg) [inline bin/btc-server/src/wallet/util.rs]: Can we check for `tx_weight.to_wu()` > 0 here to avoid 0 division ?

File: bin/btc-server/src/wallet/coin_selection.rs [hunk 5]
```diff
@@ -68,91 +84,92 @@ pub(crate) fn coin_selection(
...
-        fee_rate.fee_wu(estimated_output_weight).ok_or(CoinSelectionError::FeeRateOverflow)?;
-
-    let coin_select =
+    // Basic calculations
+    let pegouts = outputs
+        .into_iter()
+        .map(|(txout, pegout_id)| (txout, pegout_id.as_bytes()))
+        .collect::<Vec<_>>();
+    let total_utxos_value = available_utxos.values().map(|u| u.output.value).sum::<Amount>();
+    let total_pegout_target = pegouts.clone().iter().map(|(txout, _)| txout.value).sum::<Amount>();
```
Comment (rwlockbg) [inline bin/btc-server/src/wallet/coin_selection.rs]: Why do you clone in here. Cannot you just take a ref (.as_ref()) or & and rather clone the txout.value which I believe is even Copy inside the aggregator ?

File: bin/btc-server/src/wallet/coin_selection.rs [hunk 6]
```diff
@@ -68,91 +84,92 @@ pub(crate) fn coin_selection(
...
+    let (psbt, filtered_pegout_ids) = apply_fees_and_create_psbt(
+        &selected_inputs,
+        pegouts.clone(),
+        change_script.clone(),
+        fee_rate,
+    )?;
+
+    // update total pegout target to reflect the filtered pegout ids
+    let updated_pegout_target = pegouts
+        .clone()
```
Comment (rwlockbg) [inline bin/btc-server/src/wallet/coin_selection.rs]: Again not sure if we need cloning here

File: bin/btc-server/src/wallet/coin_selection.rs [hunk 7]
```diff
@@ -68,91 +84,92 @@ pub(crate) fn coin_selection(
...
+        .iter()
+        .filter(|(_, pegout_id)| !filtered_pegout_ids.contains(pegout_id))
```
Comment (rwlockbg) [inline bin/btc-server/src/wallet/coin_selection.rs]: There is a better method called `filter_map` that does both in one go.

File: bin/btc-server/src/wallet/util.rs (hunk 2 above)
Comment (DariusParvin) [reply bin/btc-server/src/wallet/util.rs]: In practice no. Something else would have to be seriously wrong for the weight to be so huge. But I guess it's better practice to always handle the overflow possibilities, at least so that it makes the code easier to reason about (and a better habit for me too :) )

File: bin/btc-server/src/wallet/coin_selection.rs (hunk 5 above)
Comment (DariusParvin) [reply bin/btc-server/src/wallet/coin_selection.rs]: thanks you yes, good catch 🙏  i am still getting used to rust

**Review by scottmillner — DISMISSED** [review]
🔥 🔥 🔥 
so much better now
easier to reason about with logic broken out into separate methods
and great test scenarios

awesome job!!

File: bin/btc-server/src/wallet/coin_selection.rs [hunk 8]
```diff
@@ -303,204 +536,279 @@ mod tests {
...
+                    TxOut {
+                        value: Amount::from_sat(value_sats),
+                        script_pubkey: random_p2wpkh_scriptpubkey(),
+                    },
+                    create_random_pegout_id(),
+                )
+            })
+            .collect();
 
         let required_utxos = HashMap::new();
```
Comment (lamafab) [inline bin/btc-server/src/wallet/coin_selection.rs:737]: Just pointing out that this is always empty. But you're not testing specifically for this anyway.

File: bin/btc-server/src/wallet/coin_selection.rs [hunk 9]
```diff
@@ -163,97 +188,267 @@ pub(crate) fn coin_selection(
...
+///
+/// **Benefits of larger change:**
+/// - More useful for future pegouts (reduces need for multiple UTXOs)
+/// - Provides UTXO consolidation over time
+///
+/// **Benefits of smaller change:**
+/// - Keeps more liquidity available while the current pegout is waiting to be confirmed
+///
+/// The target change is calculated as a percentage of the pegout value, with a ceiling
+/// to prevent excessive liquidity lockup and a floor to prevent the change from being too small.
```
Comment (lamafab) [inline bin/btc-server/src/wallet/coin_selection.rs:323]: Can you explain how the return value of this function (`target_change`) should be interpreted? You essentially take `total_pegout_target + target_change` and pass that as the `target_amount` to the coin selection algorithm. But what confuses me here is the `max_change_value`, which is 5% of the `remaining_utxo_values` (which is the _actual change_).

How does this fit into the larger picture?

File: bin/btc-server/src/wallet/coin_selection.rs [hunk 10]
```diff
@@ -163,97 +188,267 @@ pub(crate) fn coin_selection(
...
-
-            if change_output_value != total_input_value - target_amount {
-                return Err(SanityCheckError::BadRefundBalance {
-                    change_output_value,
-                    total_input_value,
-                    target_amount,
-                }
-                .into());
-            }
+    Ok(target_change)
```
Comment (lamafab) [inline bin/btc-server/src/wallet/coin_selection.rs]: You could simplify all those checks by just doing something like this - up to you:

```rust
    Ok(target_change
        .min(max_change_value)
        // for small pegouts, set the change is at least the minimum change amount
        .max(Amount::from_sat(MIN_CHANGE_SATS))
        // this is an edge case, probably only relevant for test cases,
        // if the minimum is more than the remaining utxos just return the remaining utxos value
        .min(remaining_utxos_value))
 ```

File: bin/btc-server/src/wallet/coin_selection.rs [hunk 11]
```diff
@@ -163,97 +188,267 @@ pub(crate) fn coin_selection(
...
+        return Err(CoinSelectionError::OutputsCannotBeEmpty);
+    }
+
+    let base_fee_per_output =
+        absolute_fee.checked_div(num_outputs as u64).ok_or(CoinSelectionError::FeeRateOverflow)?;
+    let remainder = absolute_fee % num_outputs as u64;
+
+    let mut fees_to_subtract = vec![base_fee_per_output; num_outputs];
+
+    // Distribute the remainder one sat at a time to highest value outputs
```
Comment (lamafab) [inline bin/btc-server/src/wallet/coin_selection.rs:365]: Very interesting way of doing this :)

**Review by lamafab — DISMISSED** [review]
Alright, looks great! I would like a little more theory on `fn calculate_target_change()`, though - see comments.

File: bin/btc-server/src/wallet/coin_selection.rs (hunk 9 above)
Comment (DariusParvin) [reply bin/btc-server/src/wallet/coin_selection.rs:323]: ah yeah i see the confusion. What you said was correct apart from the very last part: the `remaining_utxo_values` in the code refers to the _total_ remaining utxo value of all L1 funds. It comes from:

```
    let remaining_utxos_value = total_utxos_value
        .checked_sub(total_pegout_target)
        .ok_or(InsufficientFunds { needed: total_pegout_target, available: total_utxos_value })?;
```

where `total_utxos_value` is the sum of all `available_utxos`. Maybe `remaining_available_utxo_value` is clearer?

the idea for not going over 5% is to not take away too much remaining liquidity for future pegouts (while the current pegout is pending).

File: bin/btc-server/src/wallet/coin_selection.rs (hunk 10 above)
Comment (DariusParvin) [reply bin/btc-server/src/wallet/coin_selection.rs]: oh, nice! way easier to read

File: bin/btc-server/src/wallet/coin_selection.rs (hunk 9 above)
Comment (lamafab) [reply bin/btc-server/src/wallet/coin_selection.rs:323]: > refers to the _total_ remaining utxo value of all L1 funds

Ah, of course. I was thinking per transaction, for some reason. Makes sense. I think the current variable naming is good.

**Review by rwlockbg — APPROVED** [review]

**Review by scottmillner — APPROVED** [review]

**Review by lamafab — APPROVED** [review]

File: bin/btc-server/src/wallet/coin_selection.rs (hunk 8 above)
Comment (DariusParvin) [reply bin/btc-server/src/wallet/coin_selection.rs:737]: thanks for pointing this out, because there was a bug related to this part of the code that came up during the `conflicting_inputs` test fix. I've added tests covering it in the PR for that.
